File: src/org/candango/carcara/model/database/Table.java
package org.candango.carcara.model.database;

import java.util.ArrayList;

import org.candango.carcara.util.CodeHandler;

public class Table {
	
	private String name;
	
    private ArrayList<Field> fieldList = new ArrayList<Field>();
	
    public void addField( Field field ) {
    	fieldList.add( field );
    }
    
    public Field[] getFields() {
    	Field[] field = new Field[ fieldList.size() ];
    	return fieldList.toArray( field );
    }
    
    public Field[] getSerials() {
    	int serialCount = 0;
    	
    	Field[] fields = new Field[ fieldList.size() ];
    	
    	for( Field field : fieldList.toArray( fields ) )  {
    		if( field.isSerial() ) {
    			serialCount++;
    		}
    	}
    	
    	Field[] serials = new Field[ serialCount ];
    	
    	int i = 0;
    	
    	for( Field field : fieldList.toArray( fields ) )  {
    		if( field.isSerial() ) {
    			serials[ i++ ] = field;
    		}
    	}
    	
    	return serials;
    }
    
    public Field[] getNonSerials() {
    	int nonSerialCount = 0;
    	
    	Field[] fields = new Field[ fieldList.size() ];
    	
    	for( Field field : fieldList.toArray( fields ) )  {
    		if( !field.isSerial() ) {
    			nonSerialCount++;
    		}
    	}
    	
    	Field[] nonSerials = new Field[ nonSerialCount ];
    	
    	int i = 0;
    	
    	for( Field field : fieldList.toArray( fields ) )  {
    		if( !field.isSerial() ) {
    			nonSerials[ i++ ] = field;
    		}
    	}
    	
    	return nonSerials;
    }
    
    public Field[] getPks() {
    	int pkCount = 0;
    	
    	Field[] fields = new Field[ fieldList.size() ];
    	
    	for( Field field : fieldList.toArray( fields ) )  {
    		if( field.isPk() ) {
    			pkCount++;
    		}
    	}
    	
    	Field[] pks = new Field[ pkCount ];
    	
    	int i = 0;
    	
    	for( Field field : fieldList.toArray( fields ) )  {
    		if( field.isPk() ) {
    			pks[ i++ ] = field;
    		}
    	}
    	
    	return pks;
    }
    
    public Field[] getNonPks() {
    	int nonPkCount = 0;
    	
    	Field[] fields = new Field[ fieldList.size() ];
    	
    	for( Field field : fieldList.toArray( fields ) )  {
    		if( !field.isPk() ) {
    			nonPkCount++;
    		}
    	}
    	
    	Field[] nonPks = new Field[ nonPkCount ];
    	
    	int i = 0;
    	
    	for( Field field : fieldList.toArray( fields ) )  {
    		if( !field.isPk() ) {
    			nonPks[ i++ ] = field;
    		}
    	}
    	
    	return nonPks;
    }
    
    public String toString() {
    	String out = "*\tDatabase " +  getName() + "\n";
    	out += "*\t\tFields:" + "\n";
    	for( Field field : getFields() ) {
    		out += field;
    	}
    	return out;
    }

	public String getName() {
		return name;
	}

	public void setName(String name) {
		this.name = name;
	}
    
	public String getEntityName() {
		return CodeHandler.getEntityName( getName() );
	}
	
	public String getAttributeName() {
		return CodeHandler.getAttributeName( getName() );
	}
}

File: src/Carcara/Model/Database/Table.php
<?php
/**
 * Carcara (http://carcara.candango.org)
 *
 * @link      http://github.com/candango/carcara
 * @copyright Copyright (c) 2018 Flavio Garcia
 * @license   https://www.apache.org/licenses/LICENSE-2.0  Apache-2.0
 */

namespace Candango\Carcara\Model\Database
{

    use Candango\Carcara\Lexicon;

    class Table
    {
        /**
         * Table name
         *
         * @var string
         */
        private $name;

        /**
         * Table fields
         *
         * @var array
         */
        private $fields = array();

        /**
         * Table name getter
         *
         * @return string
         */
        public function getName()
        {
            return $this->name;
        }

        /**
         * Table name setter
         *
         * @param string $name
         */
        public function setName($name)
        {
            $this->name = $name;
        }

        /**
         * Add a new field to the table
         *
         * @param $field
         */
        public function addField($field)
        {
            $this->fields[] = $field;
        }

        /**
         * Table fields getter
         *
         * @return array
         */
        public function getFields()
        {
            return $this->fields;
        }

        /**
         *
         * @return array
         */
        public function getPkFields() {
            $pks = array();
            foreach ($this->fields as $field) {
                if ($field->isPk()) {
                    $pks[] = $field;
                }
            }
            return $pks;
        }

        /**
         *
         * @return array
         */
        public function getNonPkFields() {
            $nonPks = array();
            foreach ($this->fields as $field) {
                if (!$field->isPk()) {
                    $nonPks[] = $field;
                }
            }
            return $nonPks;
        }

        /**
         *
         * @return array
         */
        public function getSerialFields() {
            $serials = array();
            foreach ($this->fields as $field) {
                if ($field->isSerial()) {
                    $serials[] = $field;
                }
            }
            return $serials;
        }

        /**
         *
         * @return array
         */
        public function getNonSerialFields() {
            $nonSerials = array();
            foreach ($this->fields as $field) {
                if (!$field->isSerial()) {
                    $nonSerials[] = $field;
                }
            }
            return $nonSerials;
        }

        /**
         * Table fields setter
         *
         * @param string $fields
         */
        public function setFields($fields)
        {
            $this->fields = $fields;
        }

        /**
         * Return the table entity name
         *
         * @return string
         */
        public function getEntityName()
        {
            return Lexicon::getEntityName($this->getName());
        }
    }
}
